mmv1/third_party/terraform/services/redis/resource_redis_cluster_test.go.erb

```diff
@@ -0,0 +1,138 @@
+	t.Parallel()
+
+	name := fmt.Sprintf("tf-test-%d", acctest.RandInt(t))
+
+	acctest.VcrTest(t, resource.TestCase{
+		PreCheck:                 func() { acctest.AccTestPreCheck(t) },
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderBetaFactories(t),
+		CheckDestroy:             testAccCheckRedisClusterDestroyProducer(t),
+		Steps: []resource.TestStep{
+			{
+				// create cluster with replica count 1
+				Config: createOrUpdateRedisCluster(name /* replicaCount = */, 1 /* shardCount = */, 3, true),
+			},
+			{
+				ResourceName:      "google_redis_cluster.test",
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+			{
+				// update replica count to 2
+				Config: createOrUpdateRedisCluster(name /* replicaCount = */, 2 /* shardCount = */, 3, true),
+			},
+			{
+				ResourceName:      "google_redis_cluster.test",
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+			{
+				// clean up the resource
+				Config: createOrUpdateRedisCluster(name /* replicaCount = */, 2 /* shardCount = */, 3, false),
+			},
+		},
+	})
+}
+
+// Validate that shard count is updated for the cluster
+func TestAccRedisCluster_updateShardCount(t *testing.T) {
+	t.Parallel()
+
+	name := fmt.Sprintf("tf-test-%d", acctest.RandInt(t))
+
+	acctest.VcrTest(t, resource.TestCase{
+		PreCheck:                 func() { acctest.AccTestPreCheck(t) },
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderBetaFactories(t),
+		CheckDestroy:             testAccCheckRedisClusterDestroyProducer(t),
+		Steps: []resource.TestStep{
+			{
+				// create cluster with shard count 3
+				Config: createOrUpdateRedisCluster(name /* replicaCount = */, 1 /* shardCount = */, 3, true),
+			},
+			{
+				ResourceName:      "google_redis_cluster.test",
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+			{
+				// update shard count to 5
+				Config: createOrUpdateRedisCluster(name /* replicaCount = */, 1 /* shardCount = */, 5, true),
+			},
+			{
+				ResourceName:      "google_redis_cluster.test",
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+			{
+				// clean up the resource
+				Config: createOrUpdateRedisCluster(name /* replicaCount = */, 1 /* shardCount = */, 5, false),
+			},
+		},
+	})
+}
+
+func createOrUpdateRedisCluster(name string, replicaCount int, shardCount int, preventDestroy bool) string {
+	lifecycleBlock := ""
+	if preventDestroy {
+		lifecycleBlock = `
+		lifecycle {
+			prevent_destroy = true
+		}`
+	}
+	return fmt.Sprintf(`
+resource "google_redis_cluster" "test" {
+        provider = google-beta
+        name           = "%s"
+	replica_count = %d
+	shard_count = %d
+  region         = "us-central1"
+	psc_configs {
+			network = google_compute_network.producer_net.id
+	}
+	depends_on = [
+          google_network_connectivity_service_connection_policy.default
+        ]
+	%s
+}
+
+resource "google_network_connectivity_service_connection_policy" "default" {
```

Comment:
They should be cleaned up every run, but to make these tests more friendly to our CI system can you change the names of all the resources to be "tf-test%s" and pass in the random suffix? There are lots of examples of this in other tests.

It helps randomize the names so that if one test leaves a network or policy around it doesnt interfere with others

Reply:
ip_cidr_range = "10.1.0.0/16" under subnet is another thing that I'd need to randomize. Would it be better if I had these resources one time created in the test account and reused rather?

Reply:
No, one-time setup is a difficult pattern because it makes test environment setup difficult to replicate and difficult to run locally in a developer's own project.

Does the range used need to be randomized, or can clusters use the same range inside different networks so tests can run in parallel?

We could potentially serialize test runs if we don't have a better option

Reply:
Understood. Moved to runtime creation of all resources. 

I'm seeing this error https://yaqs.corp.google.com/eng/q/4428772363542200320 repeatedly in my tests locally  
Any idea what that means?

Also, how is deletion order of resources determined? for my case I want 
1. Cluster 
2. Policy 

- I see error in logs because policy is tried to be deleted but since cluster is still ACTIVE the deletion fails. I'm guessing it could be because of the above issue with cluster creation detection.

Reply:
Added dependent resource creation.